runtime/connectors/blob/blobdownloader.go

```diff
@@ -3,119 +3,272 @@ package blob
+	return result, nil
+}
+
+// todo :: ideally planner should take ownership of the bucket and return an iterator with next returning objectWithPlan
+func plan(ctx context.Context, bucket *blob.Bucket, opts Options, globPattern string) ([]*objectWithPlan, error) {
+	var (
+		size, fetched int64
+		matchCount    int
+	)
+	planner, err := newPlanner(opts.ExtractPolicy)
+	if err != nil {
+		return nil, err
+	}
 
 	token := blob.FirstPageToken
-	for token != nil {
-		objs, nextToken, err := bucket.ListPage(ctx, token, config.GlobPageSize, listOptions)
+	for token != nil && !planner.Done() {
+		objs, nextToken, err := bucket.ListPage(ctx, token, opts.GlobPageSize, listOptions(globPattern))
 		if err != nil {
 			return nil, err
 		}
-		token = nextToken
 
+		token = nextToken
+		fetched += int64(len(objs))
 		for _, obj := range objs {
 			if matched, _ := doublestar.Match(globPattern, obj.Key); matched {
 				size += obj.Size
 				matchCount++
-				fileNames = append(fileNames, obj.Key)
+				if !planner.Add(obj) {
+					break
+				}
 			}
 		}
-
-		fetched += int64(len(objs))
-
-		if err := validateLimits(size, matchCount, fetched, config); err != nil {
+		if err := opts.validateLimits(size, matchCount, fetched); err != nil {
 			return nil, err
 		}
 	}
 
-	if len(fileNames) == 0 {
+	items := planner.Items()
+	if len(items) == 0 {
 		return nil, fmt.Errorf("no files found for glob pattern %q", globPattern)
 	}
+	return items, nil
+}
 
-	handler.FileNames = fileNames
-	if err := handler.DownloadAll(ctx); err != nil {
-		return nil, err
+// listOptions for page listing api
+func listOptions(globPattern string) *blob.ListOptions {
+	listOptions := &blob.ListOptions{BeforeList: func(as func(interface{}) bool) error {
+		// Access storage.Query via q here.
+		var q *storage.Query
+		if as(&q) {
+			// we only need name and size, adding only required attributes to reduce data fetched
+			_ = q.SetAttrSelection([]string{"Name", "Size"})
+		}
+		return nil
+	}}
+
+	prefix, glob := doublestar.SplitPattern(globPattern)
+	if !fileutil.IsGlob(glob) {
+		// single file
+		listOptions.Prefix = globPattern
+	} else if prefix != "." {
+		listOptions.Prefix = prefix
 	}
-	return handler.LocalPaths, nil
+
+	return listOptions
 }
 
-func validateLimits(size int64, matchCount int, fetched int64, config FetchConfigs) error {
-	if size > config.GlobMaxTotalSize {
-		return fmt.Errorf("glob pattern exceeds limits: would fetch more than %d bytes", config.GlobMaxTotalSize)
+func downloadObject(ctx context.Context, bucket *blob.Bucket, objpath string, file *os.File) error {
+	rc, err := bucket.NewReader(ctx, objpath, nil)
+	if err != nil {
+		return fmt.Errorf("Object(%q).NewReader: %w", objpath, err)
 	}
-	if matchCount > config.GlobMaxObjectsMatched {
-		return fmt.Errorf("glob pattern exceeds limits: matched more than %d files", config.GlobMaxObjectsMatched)
+	defer rc.Close()
+
+	_, err = io.Copy(file, rc)
+	return err
+}
+
+type Options struct {
```

Comment:
We will have to also move ExtractPolicy and its functions to top as well since it is part of Options struct which will make the iterator functions harder to find. 
I find it better to create a different file itself for options. Please let me know if this is idiomatic in go. To be honest I find java's way of organising code better here(which would have meant this configs go into separate file if class is public).